File: Scala/src/main/scala/Stronghold/DNA.scala
package Stronghold

object DNA {
  private val Nucleotides: Vector[Char] = Vector('A', 'C', 'G', 'T')

  def countNucleotides(dna: String): Vector[Int] = {
    val counts: Map[Char, Int] = dna.groupMapReduce(identity)(_ => 1)(_ + _)
    Nucleotides.map(counts.getOrElse(_, 0))
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val dna: String = reader.next()
    val result: Vector[Int] = countNucleotides(dna)
    println(result.mkString(" "))
  }
}

File: Python/stronghold/dna.py
# Counting DNA Nucleotides
import sys

from collections import Counter

NUCLEOTIDES = ('A', 'C', 'G', 'T')


def count_nucleotides(dna):
    counter = Counter(dna)
    return [counter.get(nucleotide, 0) for nucleotide in NUCLEOTIDES]


def main():
    reader = sys.stdin
    dna = next(reader).strip()
    result = count_nucleotides(dna)
    print(' '.join(map(str, result)))


if __name__ == '__main__':
    main()
